src/main/java/org/embulk/output/kintone/KintonePageOutput.java

```diff
@@ -192,143 +186,105 @@ public void onGiveup(Exception firstException, Exception lastException)
-              upsert(records, updateKeys);
-              records.clear();
-              updateKeys.clear();
-            }
-          }
-          if (records.size() > 0) {
-            upsert(records, updateKeys);
-          }
-        });
+  private void upsertPage(Page page) {
+    List<Record> records = new ArrayList<>();
+    List<UpdateKey> updateKeys = new ArrayList<>();
+    reader.setPage(page);
+    KintoneColumnVisitor visitor =
+        new KintoneColumnVisitor(
+            reader,
+            task.getColumnOptions(),
+            task.getPreferNulls(),
+            task.getIgnoreNulls(),
+            task.getUpdateKeyName()
+                .orElseThrow(() -> new RuntimeException("unreachable"))); // Already validated
+    while (reader.nextRecord()) {
+      Record record = new Record();
+      UpdateKey updateKey = new UpdateKey();
+      visitor.setRecord(record);
+      visitor.setUpdateKey(updateKey);
+      reader.getSchema().visitColumns(visitor);
+      records.add(record);
+      updateKeys.add(updateKey);
+      if (records.size() == UPSERT_BATCH_SIZE) {
+        upsert(records, updateKeys);
+        records.clear();
+        updateKeys.clear();
+      }
+    }
+    if (!records.isEmpty()) {
+      upsert(records, updateKeys);
+    }
   }
 
-  private void upsert(ArrayList<Record> records, ArrayList<UpdateKey> updateKeys) {
+  private void upsert(List<Record> records, List<UpdateKey> updateKeys) {
     if (records.size() != updateKeys.size()) {
       throw new RuntimeException("records.size() != updateKeys.size()");
     }
-    FieldType updateKeyFieldType =
-        client.app().getFormFields(task.getAppId()).get(updateKeys.get(0).getField()).getType();
-    if (!Arrays.asList(FieldType.SINGLE_LINE_TEXT, FieldType.NUMBER).contains(updateKeyFieldType)) {
-      throw new ConfigException("The update_key must be 'SINGLE_LINE_TEXT' or 'NUMBER'.");
-    }
-
-    List<Record> existingRecords = getExistingRecordsByUpdateKey(updateKeys);
-    String updateField = updateKeys.get(0).getField();
-    List<String> existingValues =
-        existingRecords.stream()
-            .map(
-                (r) -> {
-                  switch (updateKeyFieldType) {
-                    case SINGLE_LINE_TEXT:
-                      String s = r.getSingleLineTextFieldValue(updateField);
-                      return s == null ? null : s.toString();
-                    case NUMBER:
-                      BigDecimal bd = r.getNumberFieldValue(updateField);
-                      return bd == null ? null : bd.toPlainString();
-                    default:
-                      return null;
-                  }
-                })
-            .filter(v -> v != null)
-            .collect(Collectors.toList());
-
-    ArrayList<Record> insertRecords = new ArrayList<>();
-    ArrayList<RecordForUpdate> updateRecords = new ArrayList<>();
+    List<Object> existingValues = executeWithRetry(() -> getExistingValuesByUpdateKey(updateKeys));
```

Comment:
No need to retry the entire upsert, so retry only where the client is being used.